src/Microsoft.Fx.Portability/Analysis/AnalysisEngine.cs

```diff
@@ -221,6 +224,74 @@ public IEnumerable<string> FindUnreferencedAssemblies(IEnumerable<string> unrefe
+            {
+                ImmutableDictionary<FrameworkName, IEnumerable<NuGetPackageId>> packages = null;
+                var result = _packageFinder.FindPackage(assembly, targets, out packages);
+                if (result)
+                {
+                    foreach (var target in targets)
+                    {
+                        var nuGetPackageInfo = new NuGetPackageInfo(assembly, target, packages.ContainsKey(target) ? packages[target].ToList() : new List<NuGetPackageId>());
+                        nugetPackages.Add(nuGetPackageInfo);
+                    }
+                }
+            }
+            return nugetPackages;
+        }
+
+        // Get assemblies that should be removed (nuget packages exist for all targets)
+        public List<AssemblyInfo> ComputeAssembliesToRemove(IEnumerable<AssemblyInfo> userAssemblies, IEnumerable<FrameworkName> targets, out List<NuGetPackageInfo> nugetPackagesForUserAssemblies)
+        {
+            var assembliesToRemove = new List<AssemblyInfo>();
+            nugetPackagesForUserAssemblies = new List<NuGetPackageInfo>();
+            foreach (var assembly in userAssemblies)
+            {
+                if (assembly != null)
+                {
+                    ImmutableDictionary<FrameworkName, IEnumerable<NuGetPackageId>> packages = null;
+                    var result = _packageFinder.FindPackage(assembly.AssemblyIdentity, targets, out packages);
+                    if (result)
+                    {
+                        bool supportedOnAllTargets = true;
+                        foreach (var target in targets)
+                        {
+                            var nuGetPackageInfo = new NuGetPackageInfo(assembly.AssemblyIdentity, target, packages.ContainsKey(target) ? packages[target].ToList() : new List<NuGetPackageId>());
+                            if (!nuGetPackageInfo.SupportedPackages.Any())
+                            {
+                                supportedOnAllTargets = false;
+                            }
+                            nugetPackagesForUserAssemblies.Add(nuGetPackageInfo);
+                        }
+
+                        if (assembly.SkipBinaryIfPackageExists && supportedOnAllTargets)
+                        {
+                            assembliesToRemove.Add(assembly);
+                        }
+                    }
+                }
+            }
+
+            return assembliesToRemove;
+        }
+
+        public IDictionary<MemberInfo, ICollection<AssemblyInfo>> FilterDependencies(IDictionary<MemberInfo, ICollection<AssemblyInfo>> dependencies, IEnumerable<AssemblyInfo> assembliesToRemove)
+        {
+            foreach (var dependency in dependencies)
+            {
+                var assemblies = dependency.Value;
+                foreach (var assemblyToRemove in assembliesToRemove)
+                {
+                    assemblies.Remove(assemblyToRemove);
```

Comment:
This seems potentially bad - you're modifying the input dependencies structure - is there another way or at least documenting what it will do? I'd prefer filling a new structure with it.

Reply:
Changed to create a new structure